File: problems/3000/solution.ex
defmodule Solution do
  @spec area_of_max_diagonal([[integer]]) :: integer
  def area_of_max_diagonal(dimensions) do
    {_diag, area} =
      Enum.reduce(dimensions, {0, 0}, fn [length, width], {best_diag, best_area} ->
        diag = length * length + width * width
        area = length * width
        if diag > best_diag or (diag == best_diag and area > best_area),
          do: {diag, area},
          else: {best_diag, best_area}
      end)

    area
  end
end

File: problems/3000/solution.go
func areaOfMaxDiagonal(dimensions [][]int) int {
    bestDiag, bestArea := 0, 0
    for _, d := range dimensions {
        length, width := d[0], d[1]
        diag := length*length + width*width
        area := length * width
        if diag > bestDiag || (diag == bestDiag && area > bestArea) {
            bestDiag, bestArea = diag, area
        }
    }
    return bestArea
}
